src/message/dkim.rs

```diff
@@ -0,0 +1,318 @@
+    };
+    headers.append_raw(header_name, format!("v=1; a=rsa-sha256; d={domain}; s={selector}; c={canon}; q=dns/txt; t={timestamp}; h={headers_list}; bh={body_hash}; b={signature}",domain=domain, selector=selector,canon=canon,timestamp=timestamp,headers_list=headers_list,body_hash=body_hash,signature=signature));
+    headers
+}
+
+/// Canonicalize the body of an email
+fn dkim_canonicalize_body(body: &[u8], canonicalization: DkimCanonicalizationType) -> String {
+    let body = std::str::from_utf8(body).unwrap();
+    let re = Regex::new("(\r\n)+$").unwrap();
+    match canonicalization {
+        DkimCanonicalizationType::Simple => re.replace(body, "\r\n").to_string(),
+        DkimCanonicalizationType::Relaxed => {
+            let re_double_space = Regex::new("[\\t ]+").unwrap();
+            let body = re_double_space.replace_all(body, " ").to_string();
+            let re_space_eol = Regex::new("[\t ]\r\n").unwrap();
+            let body = re_space_eol.replace_all(&body, "\r\n").to_string();
+            re.replace(&body, "\r\n").to_string()
+        }
+    }
+}
+
+/// Canonicalize the value of an header
+fn dkim_canonicalize_header_value(
+    value: &str,
+    canonicalization: DkimCanonicalizationType,
+) -> String {
+    match canonicalization {
+        DkimCanonicalizationType::Simple => value.to_string(),
+        DkimCanonicalizationType::Relaxed => {
+            let re = Regex::new("\r\n").unwrap();
+            let value = re.replace_all(value, "").to_string();
+            let re = Regex::new("[\\t ]+").unwrap();
+            format!("{}\r\n", re.replace_all(&value, " ").to_string().trim_end())
+        }
+    }
+}
+
+/// Canonicalize signed headers passed as headers_list among mail_headers using canonicalization
+fn dkim_canonicalize_headers(
+    headers_list: Vec<String>,
+    mail_headers: &Headers,
+    canonicalization: DkimCanonicalizationType,
+) -> String {
+    let mut signed_headers = Headers::new();
+    let mut signed_headers_relaxed = String::new();
+    for h in headers_list.into_iter() {
+        let h = match canonicalization {
+            DkimCanonicalizationType::Simple => h,
+            DkimCanonicalizationType::Relaxed => {
+                let mut ret = String::new();
+                ret.push_str(&h);
+                ret.to_lowercase()
+            }
+        };
+        if let Some(value) = mail_headers.get_raw(&h) {
+            match canonicalization {
+                DkimCanonicalizationType::Simple => signed_headers.append_raw(
+                    HeaderName::new_from_ascii(h).unwrap(),
+                    dkim_canonicalize_header_value(value, canonicalization),
+                ),
+                DkimCanonicalizationType::Relaxed => signed_headers_relaxed.push_str(&format!(
+                    "{}:{}",
+                    h,
+                    dkim_canonicalize_header_value(value, canonicalization)
+                )),
```

Comment:
```suggestion
                DkimCanonicalizationType::Relaxed => write!(&mut signed_headers_relaxed,
                    "{}:{}",
                    h,
                    dkim_canonicalize_header_value(value, canonicalization)
                )).unwrap(),
```

Instead of allocating a fresh string we can write directly to the buffer which will grow in chunks.